parse.go

```diff
@@ -0,0 +1,566 @@
+}
+
+func NewParser(in io.Reader, bytesToRead int64, frameChannel chan [][]int) (Parser, error) {
+	buffer := dicomio.NewDecoder(in, bytesToRead, binary.LittleEndian, dicomio.ExplicitVR)
+	p := parser{
+		decoder:      buffer,
+		frameChannel: frameChannel,
+	}
+
+	metaElems := p.ParseFileHeader()
+	if buffer.Error() != nil {
+		return nil, buffer.Error()
+	}
+	parsedElements := &DataSet{Elements: metaElems}
+	p.parsedElements = parsedElements
+	return &p, nil
+}
+
+func NewParserFromBytes(data []byte, frameChannel chan [][]int) (Parser, error) {
+	return NewParser(bytes.NewBuffer(data), int64(len(data)), frameChannel)
+}
+
+func NewParserFromFile(path string, frameChannel chan [][]int) (Parser, error) {
+	file, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	st, err := file.Stat()
+	if err != nil {
+		return nil, err
+	}
+	p, err := NewParser(file, st.Size(), frameChannel)
+	p.(*parser).file = file
+	return p, err
+}
+
+func (p *parser) Parse(options ParseOptions) (*DataSet, error) {
+	// Change the transfer syntax for the rest of the file.
+	endian, implicit, err := getTransferSyntax(p.parsedElements)
+	if err != nil {
+		return nil, err
+	}
+	p.decoder.PushTransferSyntax(endian, implicit)
+	defer p.decoder.PopTransferSyntax()
+
+	if p.file != nil {
+		defer p.file.Close()
+	}
+
+	// Read the list of elements.
+	for p.decoder.Len() > 0 {
+		startLen := p.decoder.Len()
+		elem := p.ParseNext(options)
+		if p.decoder.Len() >= startLen { // Avoid silent infinite looping.
+			panic(fmt.Sprintf("ReadElement failed to consume data: %decoder %decoder: %v", startLen, p.decoder.Len(), p.decoder.Error()))
+		}
+		if elem == endOfDataElement {
+			// element is a pixel data and was dropped by options
+			break
+		}
+		if elem == nil {
+			// Parse error.
+			continue
+		}
+		if elem.Tag == dicomtag.SpecificCharacterSet {
+			// Set the []byte -> string decoder for the rest of the
+			// file.  It's sad that SpecificCharacterSet isn't part
+			// of metadata, but is part of regular attrs, so we need
+			// to watch out for multiple occurrences of this type of
+			// elements.
+			encodingNames, err := elem.GetStrings()
+			if err != nil {
+				p.decoder.SetError(err)
+			} else {
+				// TODO(saito) SpecificCharacterSet may appear in a
+				// middle of a SQ or NA.  In such case, the charset seem
+				// to be scoped inside the SQ or NA. So we need to make
+				// the charset a stack.
+				cs, err := dicomio.ParseSpecificCharacterSet(encodingNames)
+				if err != nil {
+					p.decoder.SetError(err)
+				} else {
+					p.decoder.SetCodingSystem(cs)
+				}
+			}
+		}
+		if options.ReturnTags == nil || (options.ReturnTags != nil && tagInList(elem.Tag, options.ReturnTags)) {
+			p.parsedElements.Elements = append(p.parsedElements.Elements, elem)
+		}
+
+	}
+	return p.parsedElements, p.decoder.Error()
+}
+
+func (p *parser) ParseNext(options ParseOptions) *Element {
+	tag := readTag(p.decoder)
```

Comment:
This function came over mostly as is for the refactor from `ReadElement` but this will itself need additional refactors to hopefully make it more modular.